File: Java/src/main/java/it/unibo/ss/hangman/Game.java
package it.unibo.ss.hangman;

import java.util.HashSet;
import java.util.Optional;
import java.util.Scanner;
import java.util.Set;

public class Game {
    private final String targetWord;
    private final int maxErrors;
    private int remainingErrors;
    private final Set<Character> guessedLetters;

    public Game(String targetWord, int maxErrors) {
        this.targetWord = targetWord.toLowerCase();
        this.maxErrors = maxErrors;
        this.remainingErrors = maxErrors;
        this.guessedLetters = new HashSet<>();
    }

    public boolean guess(char letter) {
        letter = Character.toLowerCase(letter);
        this.guessedLetters.add(letter);
        if (this.targetWord.indexOf(letter) >= 0) {
            return true;
        } else {
            return false;
        }
    }

    public String grid() {
        StringBuilder result = new StringBuilder();
        String vowels = "aeiou";
        for (char c : this.targetWord.toCharArray()) {
            if (this.guessedLetters.contains(c)) {
                result.append(Character.toUpperCase(c));
            } else if (vowels.indexOf(c) != -1) {
                result.append('+');
            } else if (Character.isAlphabetic(c)) {
                result.append('-');
            } else {
                result.append(c);
            }
            result.append(' ');
        }
        result.append("/ errors left: ").append(this.remainingErrors);
        return result.toString().trim();
    }

    public char readGuess(Scanner scanner) {
        Optional<Character> guess = Optional.empty();
        while(guess.isEmpty()) {
            System.out.print("Guess a letter: ");
            String input = scanner.nextLine().trim();
            if (input.length() != 1 || !Character.isLetter(input.charAt(0))) {
                System.out.println("Please enter a single letter.");
            } else {
                char inputChar = input.charAt(0);
                if (this.guessedLetters.contains(inputChar)) {
                    System.out.println("Already guessed that letter, try another");
                } else {
                    guess = Optional.of(inputChar);
                }
            }
        }
        return guess.get();
    }

    public void play(Scanner scanner) {
        while (this.remainingErrors > 0) {
            System.out.println(grid());
            char guess = readGuess(scanner);
            if(guess(guess)) {
                boolean allGuessed = true;
                for (char c : this.targetWord.toCharArray()) {
                    if (Character.isLetter(c) && !this.guessedLetters.contains(c)) {
                        allGuessed = false;
                        break;
                    }
                }
                if (allGuessed) {
                    System.out.println("Congratulations! You've guessed the word: " + this.targetWord.toUpperCase());
                    return;
                }
            } else {
                this.remainingErrors--;
                System.out.println("Wrong guess! You have " + this.remainingErrors + " errors left.");
            }
        }
        System.out.println("Game over! The word was: " + this.targetWord.toUpperCase());
    }
}

File: Python/game.py
class Game:
    def __init__(self, target_word, max_errors):
        self.target_word = target_word.lower()
        self.max_errors = max_errors
        self.remaining_errors = max_errors
        self.guessed_letters = set()  # Track guessed letters

    def guess(self, letter):
        letter = letter.lower()
        self.guessed_letters.add(letter)
        if letter not in self.target_word:
            return False
        else:
            return True

    def grid(self):
        result = []
        vowels = 'aeiou'
        for char in self.target_word:
            if char in self.guessed_letters:
                result.append(char.upper())
            elif char in vowels:
                result.append('+')
            elif char.isalpha():
                result.append('-')
            else:
                result.append(char)
        return ' '.join(result) + ' / errors left: ' + str(self.remaining_errors)

    def read_guess(self):
        while True:
            guess = input("Guess a letter: ").strip()
            if len(guess) != 1 or not guess.isalpha():
                print("Please enter a single letter.")
            elif guess.lower() in self.guessed_letters:
                print("Already guessed that letter, try another")
            else:
                return guess.lower()


    def play(self):
        while self.remaining_errors > 0:
            print(self.grid())
            guess = self.read_guess()
            if self.guess(guess):
                if all(char in self.guessed_letters for char in self.target_word):
                    print(f"Congratulations! You've guessed the word: {self.target_word.upper()}")
                    return
            else:
                self.remaining_errors -= 1
                print(f"Wrong guess! You have {self.remaining_errors} errors left.")
        print(f"Game over! The word was: {self.target_word.upper()}")
